src/EditorFeatures/TestUtilities/Utilities/TestFixtureHelper`1.cs

```diff
@@ -0,0 +1,54 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.UnitTests
+{
+    internal sealed class TestFixtureHelper<TFixture>
+        where TFixture : class, IDisposable, new()
+    {
+        private readonly object _gate = new();
+
+        /// <summary>
+        /// Holds a weak reference to the current test fixture instance. This reference allows
+        /// <see cref="GetOrCreateFixture"/> to access and add a reference to the current test fixture if one exists,
+        /// but does not prevent the fixture from being disposed after the last reference to it is released.
+        /// </summary>
+        private ReferenceCountedDisposable<TFixture>.WeakReference _weakFixture;
+
+        /// <summary>
+        /// Gets a reference to a test fixture, or creates it if one does not already exist.
+        /// </summary>
+        /// <remarks>
+        /// <para>The resulting test fixture will not be disposed until the last referencer disposes of its reference.
+        /// It is possible for more than one test fixture to be created during the life of any single test, but only one
+        /// test fixture will be live at any given point.</para>
```

Comment:
i'm not really sure what this means.  how/when does one get multiple text fixtures during a single test?  wouldn't htat be ery weird?  My intution around a fixture is that it lives for at least one test, and can potentially be used for multiple.  THis makes it seem as if this is not the case here.

Reply:
> how/when does one get multiple text fixtures during a single test?

The obvious example would be this:

```csharp
using (GetOrCreateWorkspaceFixture())
{
}

using (GetOrCreateWorkspaceFixture())
{
}
```

The workspace fixture is cleaned up as soon as the last reference is disposed. If subsequent code in the test needs a workspace it will get a new fixture. A small subset of tests using these fixtures make multiple calls into `TestAsync` (or related) to test different cases all in one test; these would often have something like this occurring transparently.

> My intution around a fixture is that it lives for at least one test, and can potentially be used for multiple

In practice, the life of the fixture doesn't really matter. Tests need to be able to run in isolation, which means any given fixture is suitable for any single test execution. If a test has multiple scenarios within one test method, it's very similar. We eliminated the use of shared workspaces across distinct tests a couple years back when we made the MEF container so fast that sharing was no longer needed.

We don't really use test fixtures in the old sense of state sharing anymore. It's just too big/tedious a refactoring to go back and fix cases that used to do it in the past.

Reply:
i see what you mean.  thanks.